src/H5FDsubfiling.c

```diff
@@ -433,18 +434,50 @@ H5FD__copy_plist(hid_t fapl_id, hid_t *id_out_ptr)
 herr_t
 H5Pset_fapl_subfiling(hid_t fapl_id, H5FD_subfiling_config_t *fa)
 {
-    H5P_genplist_t *plist     = NULL; /* Property list pointer */
-    herr_t          ret_value = FAIL;
+    H5P_genplist_t *  plist      = NULL; /* Property list pointer */
+    hid_t             ioc_fapl   = H5I_INVALID_HID;
+    H5FD_ioc_config_t ioc_config = {
+        0,
+    };
+    H5FD_subfiling_config_t subfiling_conf = {
+        0,
+    };
+    herr_t ret_value = FAIL;
 
     FUNC_ENTER_API(FAIL)
     H5TRACE2("e", "i*!", fapl_id, fa);
 
-    HDassert(fa != NULL);
+    if (NULL == (plist = H5P_object_verify(fapl_id, H5P_FILE_ACCESS)))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a file access property list")
 
-    plist = H5P_object_verify(fapl_id, H5P_FILE_ACCESS);
+    if (fa == NULL) {
+        /* Create IOC fapl */
+        ioc_fapl = H5Pcreate(H5P_FILE_ACCESS);
+        if (H5I_INVALID_HID == ioc_fapl)
+            HGOTO_ERROR(H5E_PLIST, H5E_CANTSET, FAIL, "can't create ioc fapl")
+
+        /* Get subfiling VFD defaults */
+        if (H5Pget_fapl_subfiling(fapl_id, &subfiling_conf) < 0)
```

Comment:
Why not use fapl__get_subfiling_defaults() here?  If I read H5Pget_fapl_subfiling() correctly, it only returns the default sub-filing FAPL entry if no sub-filing FAPL entry is defined in the supplied FAPL.  In most cases, this is a distinction without a difference, but I could see it resulting in perplexing bugs now and then.

Reply:
I would need to remove the HDassert(fa) in fapl__get_subfiling_defaults

Reply:
Is not  H5Pget_fapl_subfiling calling fapl__get_subfiling_defaults. And fapl__get_subfiling_defaults expects fa to to be NULL

Reply:
As you say, fapl__get_subfiling_defaults() is called by H5Pget_fapl_subfiling(), but only if the supplied fapl doesn't contain a sub-filing fapl entry.

If I read the assert in fapl__get_subfiling_defaults correctly, it wants the fa parameter to be non-NULL (HDassert(fa);).

Reply:
Changed to use fapl__get_subfiling_defaults